conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java

```diff
@@ -0,0 +1,167 @@
+    return String.format("test:expression-type:%s:%s", expressionType, expression);
+  }
+
+  /** Returns an expected fact representing that an annotation is not relevant. */
+  public static String irrelevantAnnotation(String annotationType) {
+    return String.format("test:irrelevant-annotation:%s", annotationType);
+  }
+
+  /** Returns an expected fact representing that an annotation is not relevant. */
+  public static String sinkType(String sinkType, String sink) {
+    return String.format("test:sink-type:%s:%s", sinkType, sink);
+  }
+
+  /**
+   * Returns {@code true} if {@code fact} is a legacy {@code jspecify_nullness_mismatch} assertion.
+   */
+  public boolean isNullnessMismatch() {
+    return NULLNESS_MISMATCH.matcher(getFactText()).matches();
+  }
+
+  @Override
+  String getFactText() {
+    return fact;
+  }
+
+  /** Returns the line number in the input file where the expected fact is. */
+  long getFactLineNumber() {
+    return factLineNumber;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof ExpectedFact)) {
+      return false;
+    }
+    ExpectedFact that = (ExpectedFact) o;
+    return this.getFile().equals(that.getFile())
+        && this.getLineNumber() == that.getLineNumber()
+        && this.fact.equals(that.fact)
+        && this.factLineNumber == that.factLineNumber;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(getFile(), getLineNumber(), fact);
+  }
+
+  @Override
+  public String toString() {
+    return toStringHelper(this)
+        .add("file", getFile())
+        .add("lineNumber", getLineNumber())
+        .add("fact", fact)
+        .add("factLineNumber", factLineNumber)
+        .toString();
+  }
+
+  /** Reads {@link ExpectedFact}s from comments in a file. */
+  static class Reader {
+
+    private static final Pattern EXPECTATION_COMMENT =
+        Pattern.compile(
+            "\\s*// "
+                + ("(?<expectation>"
+                    + ASSERTION_PATTERNS.stream().map(Pattern::pattern).collect(joining("|"))
+                    + ")")
+                + "\\s*");
+
+    private final Map<Long, String> facts = new HashMap<>();
+    private long lineNumber;
+
+    /** Reads expected facts from lines in a file. */
+    ImmutableList<ExpectedFact> readExpectedFacts(Path file, List<String> lines) {
+      ImmutableList.Builder<ExpectedFact> expectedFacts = ImmutableList.builder();
+      ListIterator<String> i = lines.listIterator();
+      while (i.hasNext()) {
+        String line = i.next();
+        lineNumber = i.nextIndex();
+        Matcher matcher = EXPECTATION_COMMENT.matcher(line);
+        if (matcher.matches()) {
+          String expectation = matcher.group("expectation");
+          if (expectation != null) {
+            facts.put(lineNumber, expectation.trim());
+          }
+        } else {
+          facts.forEach(
```

Comment:
This assumes that the last line is not matching `EXPECTATION_COMMENT`, right? Otherwise, `facts` might contain things that will never be added to `expectedFacts.
A comment might help here, or an assertion before the `return` that `facts` is empty.

Reply:
I've added a TODO. The upcoming PR to deal with test names will include error handling.